File: C#/GigaChadSys/GigaChadSys-DAO/MembresiaBasicDAOImpl.cs
using GigaChadSysDBManager;
using GigaChadSysModel;
using MySql.Data.MySqlClient;

namespace GigaChadSysDAO;

public class MembresiaBasicDAOImpl : MembresiaBasicDAO
{
    public List<MembresiaBasic> ListAll()
    {
        List<MembresiaBasic> membresias = new();

        string sql = @"SELECT membresia_ID, nombrePlan, costoMantenimientoMensual, activo
                       FROM MembresiaBasic
                       WHERE activo = 1";

        try
        {
            using MySqlConnection conexion = DBManager.Instance.GetConnection();
            using MySqlCommand comando = new(sql, conexion);
            using MySqlDataReader reader = comando.ExecuteReader();

            while (reader.Read())
            {
                membresias.Add(MapearMembresiaBasic(reader));
            }
        }
        catch (Exception ex)
        {
            throw new Exception("Error al listar membresías basic.", ex);
        }

        return membresias;
    }

    public MembresiaBasic? Load(int id)
    {
        MembresiaBasic? membresia = null;

        string sql = @"SELECT membresia_ID, nombrePlan, costoMantenimientoMensual, activo
                       FROM MembresiaBasic
                       WHERE membresia_ID = @membresia_ID";

        try
        {
            using MySqlConnection conexion = DBManager.Instance.GetConnection();
            using MySqlCommand comando = new(sql, conexion);

            comando.Parameters.AddWithValue("@membresia_ID", id);

            using MySqlDataReader reader = comando.ExecuteReader();

            if (reader.Read())
            {
                membresia = MapearMembresiaBasic(reader);
            }
        }
        catch (Exception ex)
        {
            throw new Exception("Error al cargar membresía basic.", ex);
        }

        return membresia;
    }

    public MembresiaBasic Save(MembresiaBasic membresiaBasic)
    {
        membresiaBasic.Activo = true;

        string sql = @"INSERT INTO MembresiaBasic
                       (nombrePlan, costoMantenimientoMensual, activo)
                       VALUES
                       (@nombrePlan, @costoMantenimientoMensual, @activo)";

        try
        {
            using MySqlConnection conexion = DBManager.Instance.GetConnection();
            using MySqlCommand comando = new(sql, conexion);

            comando.Parameters.AddWithValue("@nombrePlan", membresiaBasic.NombrePlan);
            comando.Parameters.AddWithValue("@costoMantenimientoMensual", membresiaBasic.CostoMantenimientoMensual);
            comando.Parameters.AddWithValue("@activo", membresiaBasic.Activo);

            comando.ExecuteNonQuery();

            membresiaBasic.MembresiaId = (int)comando.LastInsertedId;
        }
        catch (Exception ex)
        {
            throw new Exception("Error al registrar membresía basic.", ex);
        }

        return membresiaBasic;
    }

    public MembresiaBasic? Update(MembresiaBasic membresiaBasic)
    {
        string sql = @"UPDATE MembresiaBasic SET
                            nombrePlan = @nombrePlan,
                            costoMantenimientoMensual = @costoMantenimientoMensual,
                            activo = @activo
                       WHERE membresia_ID = @membresia_ID";

        try
        {
            using MySqlConnection conexion = DBManager.Instance.GetConnection();
            using MySqlCommand comando = new(sql, conexion);

            comando.Parameters.AddWithValue("@nombrePlan", membresiaBasic.NombrePlan);
            comando.Parameters.AddWithValue("@costoMantenimientoMensual", membresiaBasic.CostoMantenimientoMensual);
            comando.Parameters.AddWithValue("@activo", membresiaBasic.Activo);
            comando.Parameters.AddWithValue("@membresia_ID", membresiaBasic.MembresiaId);

            int filas = comando.ExecuteNonQuery();

            return filas > 0 ? membresiaBasic : null;
        }
        catch (Exception ex)
        {
            throw new Exception("Error al actualizar membresía basic.", ex);
        }
    }

    public void Remove(MembresiaBasic membresiaBasic)
    {
        membresiaBasic.Activo = false;

        string sql = @"UPDATE MembresiaBasic
                       SET activo = @activo
                       WHERE membresia_ID = @membresia_ID";

        try
        {
            using MySqlConnection conexion = DBManager.Instance.GetConnection();
            using MySqlCommand comando = new(sql, conexion);

            comando.Parameters.AddWithValue("@activo", membresiaBasic.Activo);
            comando.Parameters.AddWithValue("@membresia_ID", membresiaBasic.MembresiaId);

            comando.ExecuteNonQuery();
        }
        catch (Exception ex)
        {
            throw new Exception("Error al eliminar membresía basic.", ex);
        }
    }

    private MembresiaBasic MapearMembresiaBasic(MySqlDataReader reader)
    {
        return new MembresiaBasic
        {
            MembresiaId = reader.GetInt32("membresia_ID"),
            NombrePlan = reader.GetString("nombrePlan"),
            CostoMantenimientoMensual = reader.GetDecimal("costoMantenimientoMensual"),
            Activo = reader.GetBoolean("activo")
        };
    }
}

File: Java/GigaChadSys-DAO/src/main/java/pe/edu/pucp/gigachadsys/dao/impl/membresias/MembresiaBasicDAOImpl.java
package pe.edu.pucp.gigachadsys.dao.impl.membresias;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;
import java.util.ArrayList;
import java.util.List;

import pe.edu.pucp.gigachadsys.dao.inter.membresias.MembresiaBasicDAO;
import pe.edu.pucp.gigachadsys.dao.manager.DBManager;
import pe.edu.pucp.gigachadsys.model.membresias.MembresiaBasic;

public class MembresiaBasicDAOImpl implements MembresiaBasicDAO {

    @Override
    public List<MembresiaBasic> listAll() {
        List<MembresiaBasic> list = new ArrayList<>();

        // NO filtrar por activo.
        // Si filtras WHERE activo = 1, nunca podrás volver a activar una membresía inactiva desde admin.
        String sql =
                "SELECT membresia_ID, nombrePlan, costoMantenimientoMensual, activo " +
                        "FROM MembresiaBasic";

        try (
                Connection connection = DBManager.getInstance().getConnection();
                PreparedStatement pstm = connection.prepareStatement(sql);
                ResultSet rs = pstm.executeQuery()
        ) {
            while (rs.next()) {
                MembresiaBasic membresia = mapearMembresiaBasic(rs);
                list.add(membresia);
            }

            return list;

        } catch (SQLException e) {
            throw new RuntimeException("Error al listar membresías Basic", e);
        }
    }

    @Override
    public MembresiaBasic load(Integer id) {
        String sql =
                "SELECT membresia_ID, nombrePlan, costoMantenimientoMensual, activo " +
                        "FROM MembresiaBasic " +
                        "WHERE membresia_ID = ?";

        try (
                Connection connection = DBManager.getInstance().getConnection();
                PreparedStatement pstm = connection.prepareStatement(sql)
        ) {
            pstm.setInt(1, id);

            try (ResultSet rs = pstm.executeQuery()) {
                if (rs.next()) {
                    return mapearMembresiaBasic(rs);
                }
            }

        } catch (SQLException e) {
            throw new RuntimeException("Error al obtener membresía Basic por ID", e);
        }

        return null;
    }

    @Override
    public MembresiaBasic save(MembresiaBasic membresia) {
        String sql =
                "INSERT INTO MembresiaBasic " +
                        "(nombrePlan, costoMantenimientoMensual, activo) " +
                        "VALUES (?, ?, ?)";

        try (
                Connection connection = DBManager.getInstance().getConnection();
                PreparedStatement pstm = connection.prepareStatement(sql, Statement.RETURN_GENERATED_KEYS)
        ) {
            pstm.setString(1, normalizarNombre(membresia.getNombre(), "Membresía Basic"));
            pstm.setDouble(2, membresia.getCostoMantenimientoMensual());
            pstm.setBoolean(3, membresia.isActiva());

            pstm.executeUpdate();

            try (ResultSet rs = pstm.getGeneratedKeys()) {
                if (rs.next()) {
                    membresia.setIdMembresia(rs.getInt(1));
                }
            }

            return membresia;

        } catch (SQLException e) {
            throw new RuntimeException("Error al registrar membresía Basic", e);
        }
    }

    @Override
    public MembresiaBasic update(MembresiaBasic membresia) {
        String sql =
                "UPDATE MembresiaBasic SET " +
                        "nombrePlan = ?, " +
                        "costoMantenimientoMensual = ?, " +
                        "activo = ? " +
                        "WHERE membresia_ID = ?";

        try (
                Connection connection = DBManager.getInstance().getConnection();
                PreparedStatement pstm = connection.prepareStatement(sql)
        ) {
            pstm.setString(1, normalizarNombre(membresia.getNombre(), "Membresía Basic"));
            pstm.setDouble(2, membresia.getCostoMantenimientoMensual());
            pstm.setBoolean(3, membresia.isActiva());
            pstm.setInt(4, membresia.getIdMembresia());

            int filas = pstm.executeUpdate();

            if (filas == 0) {
                throw new RuntimeException("No se encontró la membresía Basic con ID " + membresia.getIdMembresia());
            }

            return membresia;

        } catch (SQLException e) {
            throw new RuntimeException("Error al actualizar membresía Basic", e);
        }
    }

    @Override
    public void remove(MembresiaBasic membresia) {
        String sql =
                "UPDATE MembresiaBasic SET " +
                        "activo = 0 " +
                        "WHERE membresia_ID = ?";

        try (
                Connection connection = DBManager.getInstance().getConnection();
                PreparedStatement pstm = connection.prepareStatement(sql)
        ) {
            pstm.setInt(1, membresia.getIdMembresia());
            pstm.executeUpdate();

        } catch (SQLException e) {
            throw new RuntimeException("Error al desactivar membresía Basic", e);
        }
    }

    private MembresiaBasic mapearMembresiaBasic(ResultSet rs) throws SQLException {
        MembresiaBasic membresia = new MembresiaBasic();

        membresia.setIdMembresia(rs.getInt("membresia_ID"));
        membresia.setNombre(rs.getString("nombrePlan"));
        membresia.setCostoMantenimientoMensual(rs.getDouble("costoMantenimientoMensual"));
        membresia.setActiva(rs.getBoolean("activo"));

        return membresia;
    }

    private String normalizarNombre(String nombre, String valorPorDefecto) {
        if (nombre == null || nombre.trim().isEmpty()) {
            return valorPorDefecto;
        }

        return nombre.trim();
    }
}